src/policy/mallocspace/global.rs

```diff
@@ -63,6 +63,18 @@ impl<VM: VMBinding> SFT for MallocSpace<VM> {
         is_marked::<VM>(object, Ordering::SeqCst)
     }
 
+    fn pin_object(&self, _object: ObjectReference) -> bool {
+        false
+    }
+
+    fn unpin_object(&self, _object: ObjectReference) -> bool {
+        false
+    }
+
+    fn is_object_pinned(&self, _object: ObjectReference) -> bool {
+        false
```

Comment:
`is_object_pinned` returns `true` for immortal space and large object space. Why does it return `false` for malloc space?

Reply:
Maybe we can do this: make `pin/unpin/is_pinned` only about pinning, and additionally we change `is_movable()` to take an object reference as the argument:
https://github.com/mmtk/mmtk-core/blob/644628bcc39e649e4125c092ad325bddbd24aaaf/src/policy/sft.rs#L50

So for all the policies,
|policy|pin/unpin|is_pinned|is_movable|
|-|-|-|-|
| policy with pinning support (immix) | set/unset pin bit | check pin bit | check defrag and check pin bit |
| non moving policy (immortal/marksweep) | false | false | false |
| compulsory moving policy (copy/markcompact) | panic | false | true |

This is probably clearer.

Reply:
You don't have to make the changes about `is_movable` in this PR. If we can agree on this solution, we just need to get `pin/unpin/is_pinned` right in this PR.

Reply:
### Updated

We define `pin/unpin` as this: if a policy does not support pin/unpin objects, it should panic. Otherwise it returns a boolean to indicate whether this call to `pin` actually pinned the object (or this call to `unpin` actually unpinned the object). 
We define `is_pinned` as this: it returns true if the object is guaranteed that it won't move (pinned, or in a non-moving policy), or returns false if the object may move.

We do not change `is_movable`.

|policy|pin/unpin|is_pinned|
|-|-|-|
| policy with pinning support (immix) | set/unset pin bit | check pin bit |
| non moving policy (immortal/marksweep) | false | true |
| compulsory moving policy (copy/markcompact) | panic | false |

Reply:
There is one thing I am not sure.  By "unsuccessful", it may have two different meanings.

1.  It is unsuccessful because the object was never pinned before, or another thing (thread, or scope) has already unpinned the object before this `unpin()`.  In this case, the object is guaranteed not to be pinned after `unpin()`.  Subsequent `is_pinned()` will return false.
2.  It is unsuccessful because the space is a non-moving policy.  In this case the object is still pinned.  Subsequent `is_pinned()` still returns true.

I am not sure because I cannot think of a scenario where the user (VM binding) needs to know which case the failure is.  But no matter which case it is, I think it is always a fatal failure because bug-free program should never make an unsuccessful `unpin()` call. Again I can't think of a scenario where the VM can recover from an unsuccessful `unpin()`.

Reply:
One possibility is to change the return value of `unpin()` to `Result<(), UnpinError>` so that the return value doesn't reflect whether the object was pinned before, or will remain pinned after.  The `UnpinError` is used to indicate the concrete error.

Reply:
I can agree with this table: 

<!DOCTYPE html>

policy | pin | unpin | is_pinned
-- | -- | -- | --
policy with pinning support (immix) | set pin bit | unset pin bit | check pin bit
non moving policy (immortal/marksweep) | false | false | true
compulsory moving policy (copy/markcompact) | panic | panic | false

I think returning false when unpinning an object of a non-moving policy, makes more sense. The action of unpinning was not performed, therefore the object remains pinned in that case. For policies with pinning support, if unpinning was not successful, the object had not been pinned before (and `is_pinned` should return false afterwards). I can add that table as a comment to the documentation in the SFT methods so the semantics gets documented.
If you guys agree, I can make those changes, and we can work towards merging this.

Reply:
That looks good to me. Probably also make it clear about our reasoning for this table as comments:
* for non moving policy, all the objects are considered as forever pinned, thus attempting to pin or unpin them will not succeed and will always return false.
* for compulsory moving policy, pin/unpin is impossible, thus we will panic.